File: DTFlinkBatch/src/main/java/com/kevin/java/transformation/FlatMapDemo.java
package com.kevin.java.transformation;

import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.api.java.ExecutionEnvironment;
import org.apache.flink.api.java.operators.DataSource;
import org.apache.flink.util.Collector;

import java.util.ArrayList;

/**
 * @author caonanqing
 * @version 1.0
 * @description     flatMap：取出一个元素并产生零个，一个或多个元素
 * @createDate 2020/3/11
 */
public class FlatMapDemo {

    public static void main(String[] args) throws Exception {

        // 获取运行环境
        ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();

        ArrayList<String> list = new ArrayList<>();
        list.add("kevin java");
        list.add("coco python");
        list.add("Mr.cao scala");

        // 将数据转成DataSet算子
        DataSource<String> data = env.fromCollection(list);

        data.flatMap(new FlatMapFunction<String, String>() {
            @Override
            public void flatMap(String value, Collector<String> out) throws Exception {
                for (String s : value.split(" ")) {
                    out.collect(s);
                }
            }
        }).print();

    }
}

File: dt-flink-datastream/src/main/scala/com/kevin/scala/transform/FlatMapDemo.scala
package com.kevin.scala.transform

import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment

/**
 * @author caonanqing
 * @version 1.0
 * @description   flatMap : 输入一个元素，输出一个或多个元素
 *
 * @createDate 2020/5/16
 */
object FlatMapDemo {

  def main(args: Array[String]): Unit = {

    // 获取运行环境
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    import org.apache.flink.api.scala._

    // 创建数据集
    val dataStream = env.fromCollection(List("kevin cao","mrcao coco","mr.cao kevin"))

    // 切分元素
    val stream = dataStream.flatMap{ str => str.split(" ") }

    // 打印输出
    stream.print("stream")

    // 执行程序
    env.execute("FlatMapDemo")
  }
}
